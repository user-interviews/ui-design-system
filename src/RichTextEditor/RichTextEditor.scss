@import '../../scss/theme.scss';

.RichTextEditor {
  width: 100%;

  .RichTextEditor__character-count {
    @include synth-font-type-20;
    margin-top: 0.5rem;
  }
}


.RichTextEditor__field--error {
  .ProseMirror {
    border-color: $ux-red-600;
  }
}

.ProseMirror {
  min-height: 5rem;
  background: $ux-neutral-100;
  border: 1px solid $ux-gray-400;
  border-radius: 0 0 $ux-border-radius $ux-border-radius;
  padding: 1rem;
  transition: border-color 0.15s ease-in-out, box-shadow 0.15s ease-in-out;

  &:focus {
    border: 0.06rem solid $ux-blue-500;
    box-shadow: $input-focus-box-shadow;
    color: $input-focus-color;
    filter: drop-shadow(0px 0px 3px rgba(102, 175, 233, 0.7));
    outline: none;
  }

  > * + * {
    margin-top: 0.75em;
  }

  ul,
  ol {
    padding: 0 1rem;
  }

  li > p {
    margin: 0;
  }

  p.is-editor-empty:first-child::before {
    color: $ux-gray-700;
    content: attr(data-placeholder);
    float: left;
    height: 0;
    pointer-events: none;
  }
}
