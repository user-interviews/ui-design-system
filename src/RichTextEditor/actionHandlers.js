export const createActionHandlers = (editor) => ({
  link: () => {
    // TODO: use DS components for link prompt
    const previousUrl = editor.getAttributes('link').href;
    const url = window.prompt('URL', previousUrl);

    // cancelled
    if (url === null) {
      return;
    }

    // empty
    if (url === '') {
      editor.chain().focus().extendMarkRange('link').unsetLink()
        .run();

      return;
    }

    const protocolRegex = /^http(s?):\/\//;
    const absoluteUrl = url.match(protocolRegex) ? url : `//${url}`;

    // update link
    editor.chain().focus().extendMarkRange('link').setLink({ href: absoluteUrl })
      .run();
  },
  unlink: () => editor.chain().focus().unsetLink().run(),
  bold: () => editor.chain().focus().toggleBold().run(),
  italic: () => editor.chain().focus().toggleItalic().run(),
  unorderedList: () => editor.chain().focus().toggleBulletList().run(),
  orderedList: () => editor.chain().focus().toggleOrderedList().run(),
});
