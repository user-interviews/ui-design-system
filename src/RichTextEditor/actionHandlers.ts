import type { Editor } from '@tiptap/core';

export const createActionHandlers = (editor: Editor) => ({
  link: () => {
    // TODO: use DS components for link prompt
    const previousUrl = editor.getAttributes('link').href;
    const url = window.prompt('URL', previousUrl);

    // cancelled
    if (url === null) {
      return false;
    }

    // empty
    if (url === '') {
      editor
        .chain()
        .focus()
        .extendMarkRange('link')
        .unsetLink()
        .run();

      return false;
    }

    const protocolRegex = /^http(s?):\/\//;
    const absoluteUrl = url.match(protocolRegex) ? url : `//${url}`;

    // update link
    editor
      .chain()
      .focus()
      .extendMarkRange('link')
      .setLink({ href: absoluteUrl })
      .run();

    return false;
  },
  unlink: () => editor.chain().focus().unsetLink().run(),
  bold: () => editor.chain().focus().toggleBold().run(),
  italic: () => editor.chain().focus().toggleItalic().run(),
  unorderedList: () => editor.chain().focus().toggleBulletList().run(),
  orderedList: () => editor.chain().focus().toggleOrderedList().run(),
});
