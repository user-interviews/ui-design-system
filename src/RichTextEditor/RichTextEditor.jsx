import './RichTextEditor.scss';

import React from 'react';
import * as propTypes from 'prop-types';

import classNames from 'classnames';

import { EditorContent, useEditor } from '@tiptap/react';

import Bold from '@tiptap/extension-bold';
import BulletList from '@tiptap/extension-bullet-list';
import CharacterCount from '@tiptap/extension-character-count';
import Document from '@tiptap/extension-document';
import HardBreak from '@tiptap/extension-hard-break';
import History from '@tiptap/extension-history';
import Italic from '@tiptap/extension-italic';
import Link from '@tiptap/extension-link';
import ListItem from '@tiptap/extension-list-item';
import OrderedList from '@tiptap/extension-ordered-list';
import Paragraph from '@tiptap/extension-paragraph';
import Placeholder from '@tiptap/extension-placeholder';
import Text from '@tiptap/extension-text';

import sanitizeHtml from 'sanitize-html';

import { LoadingSkeleton } from 'src/LoadingSkeleton';
import { TemplateVariable, buildSuggestions } from './TemplateVariable';

import RichTextEditorMenuBar from './RichTextEditorMenuBar';

import { OneLineLimit } from './oneLineLimit';

import { RichTextEditorActions, RichTextEditorAllActionsArray, RichTextEditorDefaultActionsArray } from './richTextEditorActions';
import { createActionHandlers } from './actionHandlers';

const ExtendedLink = Link.extend({
  addKeyboardShortcuts() {
    const actionHandlers = createActionHandlers(this.editor);

    return {
      'Mod-k': actionHandlers.link,
    };
  },
});

const RichTextEditor = ({
  allowedAttributes,
  allowedTags,
  ariaAttributes,
  availableActions,
  characterLimit,
  className,
  hasErrors,
  id,
  initialValue,
  isOneLine,
  onChange,
  placeholder,
  templateVariables,
}) => {
  const oneLineExtension = isOneLine ? [OneLineLimit] : [];
  const hasTemplateVariables = templateVariables.length > 0;

  const requiredExtensions = [
    Document,
    Text,
    History,
    HardBreak,
    Paragraph,
    ListItem,
    Placeholder.configure({
      placeholder,
    }),
    CharacterCount.configure({
      limit: characterLimit,
    }),
  ];

  const templateVariablesExtension = hasTemplateVariables ?
    [TemplateVariable.configure({
      HTMLAttributes: {
        class: 'RichTextEditor__TemplateVariable',
      },
      suggestion: buildSuggestions(templateVariables),
    })] : [];

  const optionalExtensions = [
    {
      name: RichTextEditorActions.BOLD,
      config: Bold,
    },
    {
      name: RichTextEditorActions.ITALIC,
      config: Italic,
    },
    {
      name: RichTextEditorActions.LINK,
      config: ExtendedLink,
    },
    {
      name: RichTextEditorActions.UNORDERED_LIST,
      config: BulletList,
    },
    {
      name: RichTextEditorActions.ORDERED_LIST,
      config: OrderedList,
    },
  ].filter((extension) => availableActions.includes(extension.name))
    .map((extension) => extension.config);

  const extensions = [
    ...oneLineExtension,
    ...requiredExtensions,
    ...templateVariablesExtension,
    ...optionalExtensions,
  ];

  const editor = useEditor({
    extensions,
    content: initialValue,
    onUpdate: ({ editor: ttEditor }) => {
      const html = ttEditor.getHTML();

      // if allowAttributes or allowedTags aren't passed
      // then use defaults from sanitize-html by not passing that key in the options
      // https://github.com/apostrophecms/sanitize-html

      const options = {};

      if (allowedAttributes) {
        options.allowedAttributes = allowedAttributes;
      }

      if (allowedTags) {
        options.allowedTags = allowedTags;
      }

      // When using template variables, we need to whitelist some specific attributes so that
      // the editor can re-initialize correctly from db state
      if (hasTemplateVariables) {
        options.allowedAttributes = {
          ...(options.allowedAttributes || {}),
          span: ['class', 'data-id', 'data-type'],
        };

        if (options.allowedTags && !options.allowedTags.includes('span')) {
          options.allowedTags.push(['span']);
        }
      }

      const sanitizedHtml = sanitizeHtml(html, options);

      onChange(sanitizedHtml);
    },
  });

  return (
    editor ? (
      <div
        className="RichTextEditor"
        id={id}
      >
        {availableActions.length > 0 && (
          <RichTextEditorMenuBar
            availableActions={availableActions}
            editor={editor}
          />
        )}
        <EditorContent
          className={classNames(
            className,
            'RichTextEditor__field',
            { 'RichTextEditor__field--error': hasErrors },
            { 'RichTextEditor__field--without-menu-bar': availableActions.length === 0 },
          )}
          editor={editor}
          role="textbox"
          {...ariaAttributes}
        />
        {
          !!characterLimit && (
            <p className="RichTextEditor__character-count">
              {editor.storage.characterCount.characters()}/{characterLimit}
            </p>
          )
        }
      </div>
    ) : (
      <>
        <LoadingSkeleton height={40} />
        <LoadingSkeleton height={70} />
      </>
    )
  );
};

RichTextEditor.propTypes = {
  /**
   HTML attributes to allow while sanitizing the editor's content.
   If none provided then it uses sanitize-html's defaults
   https://github.com/apostrophecms/sanitize-html#default-options
  */
  allowedAttributes: propTypes.array,
  /**
   HTML tags to allow while sanitizing the editor's content.
   If none provided then it uses sanitize-html's defaults
   https://github.com/apostrophecms/sanitize-html#default-options
  */
  allowedTags: propTypes.array,
  /**
    ARIA attributes to pass along to the text editor field
    for accessibility purposes. The keys of the object should be valid
    ARIA attributes.
  */
  ariaAttributes: propTypes.object,
  /**
   Which actions to include in the taskbar. Available as constants.
   Current options are BOLD, ITALIC, LINK, UNLINK, UNORDERED_LIST and ORDERED_LIST
  */
  availableActions: propTypes.arrayOf(propTypes.oneOf(RichTextEditorAllActionsArray)),
  characterLimit: propTypes.number,
  className: propTypes.string,
  hasErrors: propTypes.bool,
  id: propTypes.string.isRequired,
  /**
   HTML provided to initialize the editor's content
  */
  initialValue: propTypes.string,
  isOneLine: propTypes.bool,
  placeholder: propTypes.string,
  templateVariables: propTypes.arrayOf(propTypes.string),
  /**
    Callback function to call with sanitized HTML when editor state changes
  */
  onChange: propTypes.func.isRequired,
};

RichTextEditor.defaultProps = {
  allowedAttributes: undefined,
  allowedTags: undefined,
  ariaAttributes: undefined,
  availableActions: RichTextEditorDefaultActionsArray,
  characterLimit: undefined,
  className: undefined,
  hasErrors: false,
  initialValue: undefined,
  isOneLine: false,
  placeholder: undefined,
  templateVariables: [],
};

export default RichTextEditor;
