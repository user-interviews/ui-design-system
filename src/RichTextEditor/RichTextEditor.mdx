import { ArgTypes,  Canvas } from '@storybook/blocks';
import RichTextEditor from './RichTextEditor';
import * as ComponentStories from './RichTextEditor.stories';

# RichTextEditor

<h4>
  RichTextEditor is a field element which provides a "what you see is what you get" (WYSIWYG) editor for HTML content.
</h4>

<Canvas of={ComponentStories.Default} />

### When to use
To allow users to edit text where formatting is important, such as messages to be sent via email.

### When to not use
When formatting is not important. In those cases regular form inputs are a better fit.

## Props

<ArgTypes of={RichTextEditor} />

## Stories

By default RichTextEditor includes all available formatting actions.
If you'd like to limit which options are available you can do so by passing the actions you'd like to the `availableActions` prop in an array.

<Canvas of={ComponentStories.AvailableActions} />

Provide an object where the keys are ARIA attributes to `ariaAttributes` to pass them along to the text input field for accessibility purposes.

<Canvas of={ComponentStories.ARIAAttributes} />

If you set a character limit then the length of the inputted text will be restricted accordingly, including if longer text is pasted.
A counter will also be shown below the input so you know how close you are to hitting the limit.

<Canvas of={ComponentStories.CharacterLimit} />

You are also able to limit input to one line. This can be useful in situations where a limited amount of text is idea, but formatting is important.
The `isOneLine` and `characterLimit` props can be used together in cases where you really need to restrict text length.

<Canvas of={ComponentStories.OneLine} />

RichTextEditor also has an error state to display that the input is invalid. The most common use case for this is when it is a required field and the input is empty,
but by setting `hasErrors` via your provided `onChange` callback you can check the HTML produced by the RichTextEditor for any requirements that you have.

<Canvas of={ComponentStories.Error} />

RichTextEditor can optionally take in a forwardedRef to allow content to be programmatically manipulated.

<Canvas of={ComponentStories.SetContent} />

RichTextEditor can be set to read-only mode by setting the `editable` prop to `false` (e.g., when user permissions don't allow them to edit a certain field).

<Canvas of={ComponentStories.NotEditable} />

RichTextEditor can be configured to be used in display mode only which is useful when displaying rich text when you have to do something extra with it (e.g character truncation with summary display).
Modify `displayMode` to `true`. This will automatically put the editor in not editable mode as well. You can also provide optional `onTruncate` callback that will be called if RichTextEditor truncates text.

<Canvas of={ComponentStories.DisplayMode} />

If you're using RichTextEditor with `displayMode` and `characterLimit` then your text will be truncated to plaintext if it's above the char limit.
But if your text is below the limit, then it'll still be displayed as rich text.

E.g with text above the `characterLimit`

<Canvas of={ComponentStories.DisplayModeWithCharacterLimitAbove} />

E.g with text below the `characterLimit`

<Canvas of={ComponentStories.DisplayModeWithCharacterLimitBelow} />

