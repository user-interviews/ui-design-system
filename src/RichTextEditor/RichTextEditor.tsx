import type { IOptions } from 'sanitize-html';
import type { Extension, Node as TipTapNode, Mark } from '@tiptap/core';

import './RichTextEditor.scss';

import React, {
  forwardRef, useImperativeHandle, type AriaAttributes, type ForwardedRef, useEffect,
} from 'react';

import classNames from 'classnames';

import { EditorContent, useEditor } from '@tiptap/react';

import Bold from '@tiptap/extension-bold';
import BulletList from '@tiptap/extension-bullet-list';
import CharacterCount from '@tiptap/extension-character-count';
import Document from '@tiptap/extension-document';
import HardBreak from '@tiptap/extension-hard-break';
import History from '@tiptap/extension-history';
import Italic from '@tiptap/extension-italic';
import Link from '@tiptap/extension-link';
import ListItem from '@tiptap/extension-list-item';
import OrderedList from '@tiptap/extension-ordered-list';
import Paragraph from '@tiptap/extension-paragraph';
import Placeholder from '@tiptap/extension-placeholder';
import Text from '@tiptap/extension-text';

import sanitizeHtml from 'sanitize-html';

import { LoadingSkeleton } from '../LoadingSkeleton';

import RichTextEditorMenuBar from './RichTextEditorMenuBar';

import { OneLineLimit } from './oneLineLimit';

import { RichTextEditorActions, RichTextEditorDefaultActionsArray } from './richTextEditorActions';
import { createActionHandlers } from './actionHandlers';

const ELLIPSIS = '...'

const ExtendedLink = Link.extend({
  addKeyboardShortcuts() {
    const actionHandlers = createActionHandlers(this.editor);

    return {
      'Mod-k': actionHandlers.link,
    };
  },
});

export type RichTextEditorProps = {
  /**
   HTML attributes to allow while sanitizing the editor's content.
   If none provided then it uses sanitize-html's defaults
   https://github.com/apostrophecms/sanitize-html#default-options
  */
  allowedAttributes?: IOptions['allowedAttributes'];
  /**
   HTML tags to allow while sanitizing the editor's content.
   If none provided then it uses sanitize-html's defaults
   https://github.com/apostrophecms/sanitize-html#default-options
  */
  allowedTags?: string[];
  /**
    ARIA attributes to pass along to the text editor field
    for accessibility purposes. The keys of the object should be valid
    ARIA attributes.
  */
  ariaAttributes?: AriaAttributes;
  /**
   Which actions to include in the taskbar. Available as constants.
   Current options are BOLD, ITALIC, LINK, UNLINK, UNORDERED_LIST and ORDERED_LIST
  */
  availableActions?: typeof RichTextEditorActions[keyof typeof RichTextEditorActions][];
  characterLimit?: number;
  className?: string;
  /**
   Custom extensions that provide functionality related to app features beyond
   the responsibility of the text editor itself. An example of such an extension
   would be template variables as the requirements of implementation would rely on
   other code processing them. Not yet supported extensions which relate solely
   to the text editor should be added via a pull request and used via `availableActions`.
   https://tiptap.dev/docs/editor/guide/custom-extensions
  */
  customExtensions?: (Extension | TipTapNode | Mark)[];
  displayMode?: boolean;
  editable?: boolean;
  hasErrors?: boolean;
  id: string;
  /**
   HTML provided to initialize the editor's content
  */
  initialValue?: string;
  isOneLine?: boolean;
  placeholder?: string;
  /**
    Callback function to call with sanitized HTML when editor state changes
  */
  onChange: (arg0: string) => void;
}

export type RichTextEditorRef = {
  setContent: (content: string) => void;
}

const RichTextEditor = forwardRef((
  {
    allowedAttributes,
    allowedTags,
    ariaAttributes,
    availableActions = RichTextEditorDefaultActionsArray,
    displayMode = false,
    editable = true,
    characterLimit,
    className,
    hasErrors,
    id,
    initialValue,
    isOneLine,
    onChange,
    placeholder,
    customExtensions = [],
  }: RichTextEditorProps,
  ref: ForwardedRef<RichTextEditorRef>,
) => {
  const shouldDisplayMenuBar = !displayMode && availableActions.length > 0;
  const shouldDisplayCharacterLimit = !!characterLimit && !displayMode;
  const isEditable = editable && !displayMode;
  const oneLineExtension = isOneLine ? [OneLineLimit] : [];

  const requiredExtensions = [
    Document,
    Text,
    History,
    HardBreak,
    Paragraph,
    ListItem,
    Placeholder.configure({
      placeholder,
    }),
    CharacterCount.configure({
      limit: characterLimit,
    }),
  ];

  const optionalExtensions = [
    {
      name: RichTextEditorActions.BOLD,
      config: Bold,
    },
    {
      name: RichTextEditorActions.ITALIC,
      config: Italic,
    },
    {
      name: RichTextEditorActions.LINK,
      config: ExtendedLink,
    },
    {
      name: RichTextEditorActions.UNORDERED_LIST,
      config: BulletList,
    },
    {
      name: RichTextEditorActions.ORDERED_LIST,
      config: OrderedList,
    },
  ].filter((extension) => availableActions.includes(extension.name))
    .map((extension) => extension.config);

  const editorExtensions = [
    ...requiredExtensions,
    ...optionalExtensions,
    ...oneLineExtension,
    ...customExtensions,
  ];

  const editor = useEditor({
    extensions: editorExtensions,
    content: initialValue,
    onUpdate: ({ editor: ttEditor }) => {
      const html = ttEditor.isEmpty ? '' : ttEditor.getHTML();

      // if allowAttributes or allowedTags aren't passed
      // then use defaults from sanitize-html by not passing that key in the options
      // https://github.com/apostrophecms/sanitize-html

      const options: IOptions = {};

      if (allowedAttributes) {
        options.allowedAttributes = allowedAttributes;
      }

      if (allowedTags) {
        options.allowedTags = allowedTags;
      }

      const sanitizedHtml = sanitizeHtml(html, options);

      onChange(sanitizedHtml);
    },
    editable: isEditable,
  });

  useImperativeHandle(ref, () => ({
    setContent: (content: string) => {
      editor?.commands.setContent(content);
      onChange(content);
    },
  }));

  useEffect(() => {
    if (editor) {
      editor.setEditable(isEditable);
    }
  }, [editor, isEditable]);

  useEffect(() => {
    if (
      characterLimit &&
      displayMode &&
      editor &&
      editor.storage.characterCount.characters() > characterLimit
    ) {
      const truncatedText = editor
        .getText()
        .substring(0, characterLimit - ELLIPSIS.length)
        .concat(ELLIPSIS);

      editor.commands.setContent(truncatedText);
    }
  }, [characterLimit, displayMode, editor]);

  return (
    editor ? (
      <div
        className="RichTextEditor"
        id={id}
      >
        {shouldDisplayMenuBar && (
          <RichTextEditorMenuBar
            availableActions={availableActions}
            editable={isEditable}
            editor={editor}
          />
        )}
        <EditorContent
          className={classNames(
            className,
            'RichTextEditor__field',
            { 'RichTextEditor__field--error': hasErrors },
            { 'RichTextEditor__field--without-menu-bar': availableActions.length === 0 },
            { 'RichTextEditor__field--display-mode': displayMode },
          )}
          editor={editor}
          role="textbox"
          {...ariaAttributes}
        />
        {
          shouldDisplayCharacterLimit && (
            <p className="RichTextEditor__character-count">
              {editor.storage.characterCount.characters()}/{characterLimit}
            </p>
          )
        }
      </div>
    ) : (
      <>
        <LoadingSkeleton height={40} />
        <LoadingSkeleton height={70} />
      </>
    )
  );
});

// eslint-disable-next-line import/no-default-export
export default RichTextEditor;
