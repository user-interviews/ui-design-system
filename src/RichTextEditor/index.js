import RichTextEditor from './RichTextEditor';
import { AllRteActions, DefaultRteActions, RteActions } from './richTextEditorActions';

export {
  AllRteActions,
  DefaultRteActions,
  RichTextEditor,
  RteActions,
};
