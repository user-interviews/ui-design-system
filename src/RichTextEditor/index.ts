import RichTextEditor from './RichTextEditor';
import { RichTextEditorAllActionsArray, RichTextEditorDefaultActionsArray, RichTextEditorActions } from './richTextEditorActions';

export {
  RichTextEditorAllActionsArray,
  RichTextEditorDefaultActionsArray,
  RichTextEditor,
  RichTextEditorActions,
};
