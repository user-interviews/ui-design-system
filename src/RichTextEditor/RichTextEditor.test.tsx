import { render, screen } from '@testing-library/react';
import React from 'react';

import RichTextEditor, { type RichTextEditorProps } from './RichTextEditor';

describe('<RichTextEditor />', () => {
  const Setup = (overrides: Omit<RichTextEditorProps, 'id' | 'onChange'> = {}) => (
    <RichTextEditor
      id="some-id"
      onChange={jest.fn()}
      {...overrides}
    />
  );

  it('renders snapshot', () => {
    const { asFragment } = render(<Setup />);

    expect(asFragment()).toMatchSnapshot();
  });

  describe('given an initial value', () => {
    it('deserializes value correctly', () => {
      render(<Setup initialValue="<p>hello world</p>" />);

      expect(screen.getByText('hello world')).toBeInTheDocument();
    });
  });
});
