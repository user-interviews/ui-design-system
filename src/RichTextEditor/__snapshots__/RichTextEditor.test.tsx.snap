// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RichTextEditor /> renders snapshot 1`] = `
<DocumentFragment>
  <div
    class="RichTextEditor"
    id="some-id"
  >
    <div
      class="RichTextEditorMenuBar"
    >
      <button
        aria-label="Bold"
        class="Button IconButton btn btn-transparent btn-sm"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-bold fa-fw"
          data-icon="bold"
          data-prefix="far"
          focusable="false"
          role="img"
          viewBox="0 0 384 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 56C0 42.7 10.7 32 24 32H72 88 212c68.5 0 124 55.5 124 124c0 34.7-14.3 66.2-37.3 88.7C339.7 264.9 368 307.1 368 356c0 68.5-55.5 124-124 124H88 72 24c-13.3 0-24-10.7-24-24s10.7-24 24-24H48V256 80H24C10.7 80 0 69.3 0 56zM212 232c42 0 76-34 76-76s-34-76-76-76H96V232H212zM96 280V432H244c42 0 76-34 76-76s-34-76-76-76H212 96z"
            fill="currentColor"
          />
        </svg>
      </button>
      <button
        aria-label="Italic"
        class="Button IconButton btn btn-transparent btn-sm"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-italic fa-fw"
          data-icon="italic"
          data-prefix="far"
          focusable="false"
          role="img"
          viewBox="0 0 384 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M128 56c0-13.3 10.7-24 24-24H360c13.3 0 24 10.7 24 24s-10.7 24-24 24H291.3L144.7 432H232c13.3 0 24 10.7 24 24s-10.7 24-24 24H24c-13.3 0-24-10.7-24-24s10.7-24 24-24H92.7L239.3 80H152c-13.3 0-24-10.7-24-24z"
            fill="currentColor"
          />
        </svg>
      </button>
      <button
        aria-label="Link"
        class="Button IconButton btn btn-transparent btn-sm"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-link fa-fw"
          data-icon="link"
          data-prefix="far"
          focusable="false"
          role="img"
          viewBox="0 0 640 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M580.3 267.2c56.2-56.2 56.2-147.3 0-203.5C526.8 10.2 440.9 7.3 383.9 57.2l-6.1 5.4c-10 8.7-11 23.9-2.3 33.9s23.9 11 33.9 2.3l6.1-5.4c38-33.2 95.2-31.3 130.9 4.4c37.4 37.4 37.4 98.1 0 135.6L433.1 346.6c-37.4 37.4-98.2 37.4-135.6 0c-35.7-35.7-37.6-92.9-4.4-130.9l4.7-5.4c8.7-10 7.7-25.1-2.3-33.9s-25.1-7.7-33.9 2.3l-4.7 5.4c-49.8 57-46.9 142.9 6.6 196.4c56.2 56.2 147.3 56.2 203.5 0L580.3 267.2zM59.7 244.8C3.5 301 3.5 392.1 59.7 448.2c53.6 53.6 139.5 56.4 196.5 6.5l6.1-5.4c10-8.7 11-23.9 2.3-33.9s-23.9-11-33.9-2.3l-6.1 5.4c-38 33.2-95.2 31.3-130.9-4.4c-37.4-37.4-37.4-98.1 0-135.6L207 165.4c37.4-37.4 98.1-37.4 135.6 0c35.7 35.7 37.6 92.9 4.4 130.9l-5.4 6.1c-8.7 10-7.7 25.1 2.3 33.9s25.1 7.7 33.9-2.3l5.4-6.1c49.9-57 47-142.9-6.5-196.5c-56.2-56.2-147.3-56.2-203.5 0L59.7 244.8z"
            fill="currentColor"
          />
        </svg>
      </button>
      <button
        aria-disabled="true"
        aria-label="Unlink"
        class="Button IconButton btn btn-transparent btn-sm"
        disabled=""
        type="button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-link-slash fa-fw"
          data-icon="link-slash"
          data-prefix="far"
          focusable="false"
          role="img"
          viewBox="0 0 640 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M38.8 5.1C28.4-3.1 13.3-1.2 5.1 9.2S-1.2 34.7 9.2 42.9l592 464c10.4 8.2 25.5 6.3 33.7-4.1s6.3-25.5-4.1-33.7L489.3 358.2l91-91c56.2-56.2 56.2-147.3 0-203.5C526.8 10.2 440.9 7.3 383.9 57.2l-6.1 5.4c-10 8.7-11 23.9-2.3 33.9s23.9 11 33.9 2.3l6.1-5.4c38-33.2 95.2-31.3 130.9 4.4c37.4 37.4 37.4 98.1 0 135.6l-95.1 95.1-45.5-35.7c24.2-53.3 14.1-117.9-29.3-161.3c-52.1-52.1-134.4-55.9-190.8-11.2L38.8 5.1zm186.3 146c36.9-22.3 85.6-17.6 117.4 14.3c26 26 34 63.3 23.7 96.4L225.1 151.1zM352.6 373.8c-20.2-2.7-39.7-11.7-55.2-27.3c-9.8-9.8-17-21.2-21.7-33.3l-54.2-42.7c-2.2 39.6 11.9 79.9 41.9 109.9c38.8 38.8 94.2 50.8 143.4 36l-54.2-42.7zm-236-186L59.7 244.8C3.5 301 3.5 392.1 59.7 448.2c53.6 53.6 139.5 56.4 196.5 6.5l6.1-5.4c10-8.7 11-23.9 2.3-33.9s-23.9-11-33.9-2.3l-6.1 5.4c-38 33.2-95.2 31.3-130.9-4.4c-37.4-37.4-37.4-98.1 0-135.6l60.9-60.9-38-29.9z"
            fill="currentColor"
          />
        </svg>
      </button>
      <button
        aria-label="Unordered List"
        class="Button IconButton btn btn-transparent btn-sm"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-list-ul fa-fw"
          data-icon="list-ul"
          data-prefix="far"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M64 64a32 32 0 1 0 0 64 32 32 0 1 0 0-64zm120 8c-13.3 0-24 10.7-24 24s10.7 24 24 24H488c13.3 0 24-10.7 24-24s-10.7-24-24-24H184zm0 160c-13.3 0-24 10.7-24 24s10.7 24 24 24H488c13.3 0 24-10.7 24-24s-10.7-24-24-24H184zm0 160c-13.3 0-24 10.7-24 24s10.7 24 24 24H488c13.3 0 24-10.7 24-24s-10.7-24-24-24H184zM96 256a32 32 0 1 0 -64 0 32 32 0 1 0 64 0zM64 384a32 32 0 1 0 0 64 32 32 0 1 0 0-64z"
            fill="currentColor"
          />
        </svg>
      </button>
      <button
        aria-label="Ordered List"
        class="Button IconButton btn btn-transparent btn-sm"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-list-ol fa-fw"
          data-icon="list-ol"
          data-prefix="far"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M24 56c0-13.3 10.7-24 24-24H80c13.3 0 24 10.7 24 24V176h16c13.3 0 24 10.7 24 24s-10.7 24-24 24H40c-13.3 0-24-10.7-24-24s10.7-24 24-24H56V80H48C34.7 80 24 69.3 24 56zM86.7 341.2c-6.5-7.4-18.3-6.9-24 1.2L51.5 357.9c-7.7 10.8-22.7 13.3-33.5 5.6s-13.3-22.7-5.6-33.5l11.1-15.6c23.7-33.2 72.3-35.6 99.2-4.9c21.3 24.4 20.8 60.9-1.1 84.7L86.8 432H120c13.3 0 24 10.7 24 24s-10.7 24-24 24H32c-9.5 0-18.2-5.6-22-14.4s-2.1-18.9 4.3-25.9l72-78c5.3-5.8 5.4-14.6 .3-20.5zM216 72H488c13.3 0 24 10.7 24 24s-10.7 24-24 24H216c-13.3 0-24-10.7-24-24s10.7-24 24-24zm0 160H488c13.3 0 24 10.7 24 24s-10.7 24-24 24H216c-13.3 0-24-10.7-24-24s10.7-24 24-24zm0 160H488c13.3 0 24 10.7 24 24s-10.7 24-24 24H216c-13.3 0-24-10.7-24-24s10.7-24 24-24z"
            fill="currentColor"
          />
        </svg>
      </button>
    </div>
    <div
      class="RichTextEditor__field"
      role="textbox"
    >
      <div
        class="tiptap ProseMirror"
        contenteditable="true"
        tabindex="0"
        translate="no"
      >
        <p
          class="is-empty is-editor-empty"
        >
          <br
            class="ProseMirror-trailingBreak"
          />
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
